File: recover-binary-search-tree/recover-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def checkForLimit(self, root, lower, upper):
        if root == None:
            return
        if root.val>lower.val and root.val<upper.val:
            self.checkForLimit(root.left, lower, root)
            self.checkForLimit(root.right, root, upper)
        else:
            if root.val<lower.val:
                root.val, lower.val = lower.val, root.val
            elif root.val>upper.val:
                root.val, upper.val = upper.val, root.val
            self.checkForLimit(root, lower, upper)
            return
    def recoverTree(self, root: Optional[TreeNode]):
        for _ in range(10):
            self.checkForLimit(root, TreeNode(-2**32), TreeNode(2**32))
        
        

File: recover-binary-search-tree/recover-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode prev;
    TreeNode violation1;
    TreeNode violation2;
    public void inorder(TreeNode root) {
        if(root == null)
            return;
        
        inorder(root.left);
        
        if(prev != null){
            if(prev.val > root.val){
                if(violation1 == null)
                    violation1 = prev;
                violation2 = root;
                }
        }
        
        prev = root;
        inorder(root.right);
    }

    public void recoverTree(TreeNode root) {
        inorder(root);
        int temp = violation1.val;
        
        violation1.val = violation2.val;
        violation2.val = temp;
    }
}